core/market/decentralized/src/db/dao/events.rs

```diff
@@ -0,0 +1,118 @@
+use diesel::{ExpressionMethods, QueryDsl, RunQueryDsl};
+use thiserror::Error;
+
+use ya_persistence::executor::Error as DbError;
+use ya_persistence::executor::{do_with_transaction, AsDao, PoolType};
+
+use crate::db::dao::demand::{demand_status, DemandState};
+use crate::db::models::MarketEvent;
+use crate::db::models::{OwnerType, Proposal, SubscriptionId};
+use crate::db::schema::market_requestor_event::dsl as dsl_requestor;
+use crate::db::DbResult;
+
+#[derive(Error, Debug)]
+pub enum TakeEventsError {
+    #[error("Removed different number of events '{num_removed}' that expected '{to_remove}'.")]
+    EventsRemovalError {
+        num_removed: usize,
+        to_remove: usize,
+    },
+    #[error("Subscription [{0}] not found. Could be unsubscribed.")]
+    SubscriptionNotFound(SubscriptionId),
+    #[error("Subscription [{0}] expired.")]
+    SubscriptionExpired(SubscriptionId),
+    #[error(transparent)]
+    DatabaseError(DbError),
+}
+
+pub struct EventsDao<'c> {
+    pool: &'c PoolType,
+}
+
+impl<'c> AsDao<'c> for EventsDao<'c> {
+    fn as_dao(pool: &'c PoolType) -> Self {
+        Self { pool }
+    }
+}
+
+impl<'c> EventsDao<'c> {
+    pub async fn add_requestor_event(&self, proposal: Proposal) -> DbResult<()> {
+        do_with_transaction(self.pool, move |conn| {
+            let event = MarketEvent::from_proposal(&proposal, OwnerType::Requestor);
+            diesel::insert_into(dsl_requestor::market_requestor_event)
+                .values(event)
+                .execute(conn)?;
+            Ok(())
+        })
+        .await
+    }
+
+    pub async fn take_requestor_events(
+        &self,
+        subscription_id: &SubscriptionId,
+        max_events: i32,
+    ) -> Result<Vec<MarketEvent>, TakeEventsError> {
+        let subscription_id = subscription_id.clone();
+        Ok(do_with_transaction(self.pool, move |conn| {
+            match demand_status(conn, &subscription_id)? {
```

Comment:
We need to do it in single transaction.
Why are you worried about mixing Dao?
That's like saying that you are worried SQL queries with multiple tables.
But of course I don't like this place, because this way Broker needs Demand existence in database, what is implicit dependency :(